empiric-package/empiric/publisher/async_fetch/bitstamp.py

```diff
@@ -0,0 +1,61 @@
+import asyncio
+import logging
+from typing import List, Union
+
+from aiohttp import ClientSession
+from empiric.core.entry import Entry
+from empiric.core.utils import currency_pair_to_key
+from empiric.publisher.assets import EmpiricAsset, EmpiricSpotAsset
+from empiric.publisher.base import PublisherFetchError, PublisherInterfaceT
+
+logger = logging.getLogger(__name__)
+
+
+class BitstampFetcher(PublisherInterfaceT):
+    BASE_URL: str = "https://www.bitstamp.net/api/v2/ticker"
+    SOURCE: str = "bitstamp"
+    publisher: str
+
+    def __init__(self, assets: List[EmpiricAsset], publisher):
+        self.assets = assets
+        self.publisher = publisher
+
+    async def _fetch_pair(
+        self, asset: EmpiricSpotAsset, session: ClientSession
+    ) -> Union[Entry, PublisherFetchError]:
+        pair = asset["pair"]
+        url = f"{self.BASE_URL}/{pair[0].lower()}{pair[1].lower()}"
+        async with session.get(url) as resp:
+            if resp.status == 404:
+                logger.debug(f"No data found for {'/'.join(pair)} from Bitstamp")
+                return PublisherFetchError(pair)
```

Comment:
yeah i'm mimicking a Result type that returns `Result<Entity, PublisherFetchError>`.  I would imagine we could filter based on the results.  Throwing might not be good because that will error the entire data pull if a single value is not found.

Reply:
In general I think throwing is an antipattern because it does not give the user the ability to handle these errors individually.  A user might have one error and 20 successes when they fetch assets, and instead of having the whole thing fail and raise an exception, we could provide the 19 successes and the 1 error and they could determine how to handle them.  otherwise they would need to catch the exception, parse for the error asset, filter the asset array, and rerun.

this is a library that provides a more rust-like Result type for python if we wanted to consider something a bit clearer.
https://github.com/rustedpy/result